### PR TITLE
LibWeb: (totally legit) CSS Animations

### PR DESCRIPTION
**Pull request by alimpfard**
Depends on #19037.

This partially implements CSS-Animations-1 (though there are references
to CSS-Animations-2).
Current limitations:
- Multi-selector keyframes are not supported.
- Most animation properties are ignored.
- Timing functions are not applied.
- Non-absolute values are not interpolated unless the target is also of
  the same non-absolute type (e.g. 10% -> 25%, but not 10% -> 20px).
- The JavaScript interface is left as an exercise for the next poor soul
  looking at this code.

With those said, this commit implements:
- Interpolation for most common types
- Proper keyframe resolution (including the synthetic from-keyframe
  containing the initial state)
- Properly driven animations, and proper style invalidation

### CODE REVIEW

File: Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp [hunk 1]
```diff
@@ -3683,7 +3683,7 @@ ErrorOr<RefPtr<StyleValue>> Parser::parse_dimension_value(ComponentValue const&
     // 2) It's a 0.
     // We handle case 1 here. Case 2 is handled by NumericStyleValue pretending to be a LengthStyleValue if it is 0.
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:3798]: Ah, this comment says I missed this when I changed parsing. We no longer do the pretending. Apparently no layout tests happened to rely on this. 🤔

**Comment by AtkinsSJ** [comment]
This is crashing for me when I visit `css-animations.html` in Ladybird.

```
Thread 1 "WebContent" received signal SIGSEGV, Segmentation fault.
0x00007f24a7af188b in ref () at /home/sam/serenity/Meta/Lagom/../../AK/RefCounted.h:28
28              ++m_ref_count;
(gdb) bt
#0  0x00007f24a7af188b in ref () at /home/sam/serenity/Meta/Lagom/../../AK/RefCounted.h:28
#1  NonnullRefPtr () at /home/sam/serenity/Meta/Lagom/../../AK/NonnullRefPtr.h:84
#2  copy_ () at /home/sam/serenity/Meta/Lagom/../../AK/Variant.h:68
#3  copy_ () at /home/sam/serenity/Meta/Lagom/../../AK/Variant.h:70
#4  copy_ () at /home/sam/serenity/Meta/Lagom/../../AK/Variant.h:70
#5  Variant () at /home/sam/serenity/Meta/Lagom/../../AK/Variant.h:305
#6  Array () at /home/sam/serenity/Meta/Lagom/../../AK/Array.h:28
#7  ResolvedKeyFrame () at /home/sam/serenity/Meta/Lagom/../../Userland/Libraries/LibWeb/CSS/StyleComputer.h:137
#8  operator() () at /home/sam/serenity/Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:2043
#9  call () at /home/sam/serenity/Meta/Lagom/../../AK/Function.h:173
#10 0x00007f24a7a5f229 in operator() () at /home/sam/serenity/Meta/Lagom/../../AK/Function.h:111
#11 for_each_effective_keyframes_at_rule () at /home/sam/serenity/Userland/Libraries/LibWeb/CSS/CSSRuleList.cpp:175
#12 0x00007f24a7af11d6 in operator()<Web::CSS::CSSStyleSheet> () at /home/sam/serenity/Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:2029
#13 0x00007f24a7aef29a in for_each_stylesheet<(lambda at /home/sam/serenity/Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:1979:41)> ()
    at /home/sam/serenity/Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:197
#14 make_rule_cache_for_cascade_origin () at /home/sam/serenity/Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:1979
#15 0x00007f24a7aeeff4 in build_rule_cache () at /home/sam/serenity/Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:2135
#16 0x00007f24a7aeece3 in build_rule_cache_if_needed () at /home/sam/serenity/Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:1965
#17 compute_style_impl () at /home/sam/serenity/Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:1906
#18 0x00007f24a7aeec29 in compute_style () at /home/sam/serenity/Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:1895
#19 0x00007f24a7b50f51 in recompute_style () at /home/sam/serenity/Userland/Libraries/LibWeb/DOM/Element.cpp:446
#20 0x00007f24a7b36fa8 in update_style_recursively () at /home/sam/serenity/Userland/Libraries/LibWeb/DOM/Document.cpp:915
#21 0x00007f24a7b37048 in operator()<Web::DOM::Node> () at /home/sam/serenity/Userland/Libraries/LibWeb/DOM/Document.cpp:928
#22 for_each_child<(lambda at /home/sam/serenity/Userland/Libraries/LibWeb/DOM/Document.cpp:926:29)> ()
    at /home/sam/serenity/Meta/Lagom/../../Userland/Libraries/LibWeb/DOM/Node.h:535
#23 update_style_recursively () at /home/sam/serenity/Userland/Libraries/LibWeb/DOM/Document.cpp:926
#24 0x00007f24a7b36d16 in update_style () at /home/sam/serenity/Userland/Libraries/LibWeb/DOM/Document.cpp:950
#25 0x00007f24a7dcaa20 in operator() () at /home/sam/serenity/Meta/Lagom/../../Userland/Libraries/LibJS/SafeFunction.h:81
#26 operator() () at /home/sam/serenity/Userland/Libraries/LibWeb/Platform/TimerSerenity.cpp:23
#27 call () at /home/sam/serenity/Meta/Lagom/../../AK/Function.h:173
#28 0x00007f24a5c837a6 in operator() () at /home/sam/serenity/Meta/Lagom/../../AK/Function.h:111
#29 timer_event () at /home/sam/serenity/Userland/Libraries/LibCore/Timer.cpp:78
#30 0x00007f24a5c6dca9 in dispatch_event () at /home/sam/serenity/Userland/Libraries/LibCore/Object.cpp:214
#31 0x000055d0167ff05e in qt_timer_fired () at /home/sam/serenity/Ladybird/EventLoopImplementationQt.cpp:91
#32 operator() () at /home/sam/serenity/Ladybird/EventLoopImplementationQt.cpp:106
#33 call () at /usr/include/x86_64-linux-gnu/qt6/QtCore/qobjectdefs_impl.h:146
#34 call<QtPrivate::List<>, void> () at /usr/include/x86_64-linux-gnu/qt6/QtCore/qobjectdefs_impl.h:256
#35 impl () at /usr/include/x86_64-linux-gnu/qt6/QtCore/qobjectdefs_impl.h:420
#36 0x00007f24a686a023 in ?? () from /lib/x86_64-linux-gnu/libQt6Core.so.6
#37 0x00007f24a68784be in QTimer::timeout(QTimer::QPrivateSignal) () from /lib/x86_64-linux-gnu/libQt6Core.so.6
#38 0x00007f24a685e62f in QObject::event(QEvent*) () from /lib/x86_64-linux-gnu/libQt6Core.so.6
#39 0x00007f24a6811a58 in QCoreApplication::notifyInternal2(QObject*, QEvent*) () from /lib/x86_64-linux-gnu/libQt6Core.so.6
#40 0x00007f24a697a203 in QTimerInfoList::activateTimers() () from /lib/x86_64-linux-gnu/libQt6Core.so.6
#41 0x00007f24a6a39bfc in ?? () from /lib/x86_64-linux-gnu/libQt6Core.so.6
```

File: Userland/Libraries/LibWeb/CSS/StyleComputer.cpp [hunk 2]
```diff
@@ -972,6 +1273,9 @@ ErrorOr<void> StyleComputer::compute_cascaded_values(StyleProperties& style, DOM
         did_match_any_pseudo_element_rules = true;
     }
 
+    // matching_rule_set.animation_rules = collect_matching_rules(element, CascadeOrigin::Animation, pseudo_element);
+    // sort_matching_rules(matching_rule_set.animation_rules);
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:1277]: Is this meant to be a FIXME? Or did you just leave it here by mistake?

File: Userland/Libraries/LibWeb/Forward.h [hunk 3]
```diff
@@ -70,10 +70,13 @@ class BorderRadiusStyleValue;
...
 class Clip;
+class CSSConditionRule;
+class CSSKeyframeRule;
+class CSSKeyframesRule;
 class ColorStyleValue;
+class CompositeStyleValue;
 class ConicGradientStyleValue;
 class ContentStyleValue;
-class CSSConditionRule;
 class CSSFontFaceRule;
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/Forward.h:80]: Nit: This sorting confuses me.

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
Some stuff. My brain melted a bit at the end there so I haven't really reviewed the "actually do animationy stuff" parts.

File: Userland/Libraries/LibWeb/CSS/StyleComputer.h [hunk 4]
```diff
@@ -145,6 +164,67 @@ class StyleComputer {
...
+            Reverse,
+            AlternateReverse,
+        } direction;
+        enum FillMode {
+            None,
+            Forward,
+            Backward,
+            Both,
+        } fill_mode;
+        DOM::Element* owning_element;
```
Comment (awesomekling) [inline Userland/Libraries/LibWeb/CSS/StyleComputer.h:205]: Let's at least use `WeakPtr` for this :)

**Comment by awesomekling** [comment]
The use of `Element` and `CSSStyleDeclaration` raw pointers looks sketchy.

File: Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp [hunk 5]
```diff
@@ -3163,6 +3166,117 @@ CSSRule* Parser::convert_to_rule(NonnullRefPtr<Rule> rule)
...
+                if (!child_tokens.has_next_token())
+                    break;
+
+                child_tokens.skip_whitespace();
+                auto token = child_tokens.next_token();
+                if (token.is_block()) {
+                    auto block_tokens = token.block().values();
+                    block_tokens.prepend(Token::create_curly(Token::Type::OpenCurly));
+                    block_tokens.prepend(Token::create_delim("*"_short_string));
+                    block_tokens.append(Token::create_curly(Token::Type::CloseCurly));
```
[the remark this answers is not in the document]
Comment (alimpfard) [reply Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:3258]: It's not immediately obvious to me how this is supposed to work, I want a CSSRuleList, not a CSSStyleDeclaration?

File: Userland/Libraries/LibWeb/CSS/StyleComputer.cpp (hunk 2 above)
Comment (alimpfard) [reply Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:1277]: Leftover from previous impl, will remove.

**Comment by alimpfard** [comment]
> The use of Element and CSSStyleDeclaration raw pointers looks sketchy

The ones in `AnimationKey`? They're just used together as a key, not dereferenced and not used otherwise.

if you mean `owning_element`, what else are we supposed to hold? JS::GCPtr is effectively an annotation type, if this is sketchy, everything else in StyleComputer is also sketchy :yakfused:

File: Userland/Libraries/LibWeb/Forward.h (hunk 3 above)
Comment (alimpfard) [reply Userland/Libraries/LibWeb/Forward.h:80]: Yeah I suppose I should move all the CSS* to the top, my bad.

**Comment by alimpfard** [comment]
I've replaced the one theoretically-unsafe `Element*` with a `JS::NonnullGCPtr`, and left a FIXME for removing the animation if its corresponding element isn't around.

File: Userland/Libraries/LibWeb/CSS/CSSRuleList.cpp [hunk 6]
```diff
@@ -141,6 +142,38 @@ void CSSRuleList::for_each_effective_style_rule(Function<void(CSSStyleRule const
         case CSSRule::Type::Supports:
             static_cast<CSSSupportsRule const&>(*rule).for_each_effective_style_rule(callback);
             break;
+        case CSSRule::Type::Keyframe:
+        case CSSRule::Type::Keyframes:
+            break;
```
Comment (konradekk) [inline Userland/Libraries/LibWeb/CSS/CSSRuleList.cpp:147]: a `default` case with a debug log here too? 🤔

(also in other similar places, optionally: a `VERIFY`?)

File: Userland/Libraries/LibWeb/CSS/Properties.json [hunk 7]
```diff
@@ -30,6 +30,103 @@
       "align-self"
     ]
   },
+  "animation": {
+    "affects-layout": true,
```
Comment (konradekk) [inline Userland/Libraries/LibWeb/CSS/Properties.json:34]: (hmm… this depends on what exactly is animated… 🤔)

File: Userland/Libraries/LibWeb/CSS/StyleComputer.cpp [hunk 8]
```diff
@@ -953,6 +966,319 @@ static ErrorOr<void> cascade_custom_properties(DOM::Element& element, Optional<C
...
+
+bool StyleComputer::Animation::is_done() const
+{
+    return progress.as_fraction() >= 0.9999f && iteration_count.has_value() && iteration_count.value() == 0;
+}
+
+void StyleComputer::ensure_animation_timer() const
+{
+    if (!m_animation_driver_timer) {
+        m_animation_driver_timer = Platform::Timer::create_repeating(16, [this] {
```
Comment (konradekk) [inline Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:1214]: i remember that `16` is assuming 60fps — not necessarily obvious for other people and maybe weʼd like to control that in the future — care to extract this somewhere, at least  as a constant?

(same for `16` below)

File: Userland/Libraries/LibWeb/CSS/StyleComputer.cpp [hunk 9]
```diff
@@ -953,6 +966,319 @@ static ErrorOr<void> cascade_custom_properties(DOM::Element& element, Optional<C
...
+        if (active_state_if_fill_forward)
+            active_state_if_fill_forward->state[to_underlying(property_id)] = next_value;
+    }
+
+    return {};
+}
+
+bool StyleComputer::Animation::is_done() const
+{
+    return progress.as_fraction() >= 0.9999f && iteration_count.has_value() && iteration_count.value() == 0;
```
Comment (konradekk) [inline Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:1208]: there was a different place with three 9s previously — looks like a constant we could unify…? (some kind of `_threshold` or `_epsilon` as people tend to name it?)

**Review by konradekk — COMMENTED** [review]
minors, as always here… 😅

File: Userland/Libraries/LibWeb/CSS/CSSRuleList.cpp (hunk 6 above)
Comment (alimpfard) [reply Userland/Libraries/LibWeb/CSS/CSSRuleList.cpp:147]: No. Not having a default makes the compiler tell you where you need to handle the new enum value.

File: Userland/Libraries/LibWeb/CSS/Properties.json (hunk 7 above)
Comment (alimpfard) [reply Userland/Libraries/LibWeb/CSS/Properties.json:34]: Not exactly, changing the animation property itself will never affect layout, the animation _running_ can of course, affect layout by changing properties that do.

File: Userland/Libraries/LibWeb/CSS/StyleComputer.cpp (hunk 8 above)
Comment (alimpfard) [reply Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:1234]: sure.

File: Userland/Libraries/LibWeb/CSS/StyleComputer.cpp (hunk 9 above)
Comment (alimpfard) [reply Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:1228]: Unless it shows up as a broken animation, the actual number here doesn't really matter (you'd only observe this on say, `animation-duration: 10000s;` where 99.99% is 1 second, but even then, that error is fine imo for an initial impl, the "real" spec-compliant implementation of this function has other parameters anyway)

File: Userland/Libraries/LibWeb/CSS/Parser/Token.h [hunk 10]
```diff
@@ -175,6 +175,22 @@ class Token {
...
+    }
+
+    static Token create_delim(String text)
+    {
+        Token token;
+        token.m_type = Type::Delim;
+        token.m_value = text;
+        return token;
+    }
+
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/CSS/Parser/Token.h:193]: These now aren't used anywhere.

File: Userland/Libraries/LibWeb/CSS/StyleComputer.cpp [hunk 11]
```diff
@@ -1365,13 +1383,26 @@ ErrorOr<void> StyleComputer::compute_cascaded_values(StyleProperties& style, DOM
...
+                        auto fill_mode_name = fill_mode_value->to_identifier();
+                        if (fill_mode_name == ValueID::Forwards)
+                            fill_mode = Animation::FillMode::Forward;
+                        if (fill_mode_name == ValueID::Backwards)
+                            fill_mode = Animation::FillMode::Backward;
+                        if (fill_mode_name == ValueID::None)
+                            fill_mode = Animation::FillMode::None;
+                        if (fill_mode_name == ValueID::Both)
+                            fill_mode = Animation::FillMode::Both;
+                    }
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:1397]: Adding the `animation-fill-mode` values to `Enums.json` would get you the to/from-identifier code for free.

**Review by AtkinsSJ — COMMENTED** [review]
Just minor things. (Accidentally clicked the wrong button for the unused token-makers comment. :sweat_smile:)

These aren't blockers, just if you're happy to make the changes right away go for it, otherwise they can be made after merging.

File: Userland/Libraries/LibWeb/CSS/StyleComputer.h [hunk 12]
```diff
@@ -126,17 +133,29 @@ class StyleComputer {
 
     JS::NonnullGCPtr<DOM::Document> m_document;
 
+    struct AnimationKeyFrameSet {
+        struct ResolvedKeyFrame {
+            struct UseInitial { };
+            Array<Variant<Empty, UseInitial, NonnullRefPtr<StyleValue const>>, to_underlying(last_property_id) + 1> resolved_properties;
```
Comment (Lubrsi) [inline Userland/Libraries/LibWeb/CSS/StyleComputer.h:139]: This is uninitialized. Fixes https://github.com/SerenityOS/serenity/pull/19050#issuecomment-1564384089
```suggestion
            Array<Variant<Empty, UseInitial, NonnullRefPtr<StyleValue const>>, to_underlying(last_property_id) + 1> resolved_properties {};
```

**Review by Lubrsi — CHANGES_REQUESTED** [review]

**Comment by alimpfard** [comment]
Changelog:
- Added `animation-direction` because I felt like it
- Fixed the uninitialised array issue @Lubrsi mentioned
- Switched to using Enums.json for `animation-{direction,fill-mode}` values
- Removed the unused Token::functions.

**Comment by awesomekling** [comment]
Very cool! Let's bring it in :)

File: Userland/Libraries/LibWeb/CSS/CSSRuleList.cpp (hunk 6 above)
Comment (konradekk) [reply Userland/Libraries/LibWeb/CSS/CSSRuleList.cpp:147]: neat! (missed that feature!)

File: Userland/Libraries/LibWeb/CSS/Properties.json (hunk 7 above)
Comment (konradekk) [reply Userland/Libraries/LibWeb/CSS/Properties.json:34]: hah! sounds reasonable!

**Pull request merged**